library(data.table)
library(dplyr)
library(tidyr)
library(stringr)
library(glue)
library(readr)
library(rebus)
library(fuzzyjoin)
library(hablar)


#Create List of Files in the Directory

interpro_files <- list.files(path = "/Users/ryanmckeown/Documents/andersen_lab/data_exploration/del_explorers/dauer_pathway/interpro",
           pattern = "tsv"

           )
# interpro_edit <- list.files(path = "/Users/ryanmckeown/Documents/andersen_lab/data_exploration/interpro_transcripts",
#                             pattern = "edited.tsv" )



gene_files <- list.files(path = "/Users/ryanmckeown/Documents/andersen_lab/data_exploration/del_explorers/dauer_pathway/interpro",
                         pattern = "csv"

)


#Read in and Format Interpro files
load_interpro <- function(file){


  inter_pro_cols <- c("ISOFORM", "MD5", "SEQUENCE_LEN", 'ANALYSIS',
                      "SIGNATURE_ID", "SIGNATURE", "START", "STOP",
                      "SCORE", "STATUS", "DATE", "ANNOTATION_ID",
                      "ANNOTATION", "GO", "PATHWAY")

  file_read <- read_tsv(file,  col_names = inter_pro_cols) %>%
    mutate(ISOFORM = toupper(.$ISOFORM)) %>% #Make Transcript Uppercase
    convert(int(START))

  #Filter to Specific InterPro Analyses

  #Filter Interpro Data to contain Analyses you are interested in
  select_interpro_analysis <- function(interpro_data, analysis){
    reduced <- filter(interpro_data, ANALYSIS %in% analysis)
    return(reduced)
  }

  reduced <- select_interpro_analysis(file_read, c("ProSiteProfiles",
                                                   "ProSitePatterns",
                                                   "CDD",
                                                   "SMART",
                                                   "Pfam",
                                                   "PRINTS"
                                                   )) %>%
    select(ISOFORM, START, STOP, SIGNATURE, SCORE , ANNOTATION, ANNOTATION_ID) #Filter For Key Columns



  transcript_name <- toupper(unique(file_read$ISOFORM))


  assign(glue("{transcript_name}"), reduced , pos = 1 ) #Load into ENV with transcript name

  #Add step to make sure its in the right case

}

#Problem is resolved if you manually delete the reactome data
# Data generated by SUPERFAMILY & GENE 3D
#Need to have some flag for when files need to be altered manually


for (f in interpro_files){
  load_interpro(f)
}

# for(f in interpro_edit){
#   load_interpro(f)
#   }

#Need to fix formatting error for Y55D5A.5D.1

# Y55D5A.5D <- Y55D5A.5D.1


#need a way to get just transcript name for DF I am analyzing

#Fails with transcripts with one Isoform (ex Daf-18)
gene_interpro <- function(isoform_root){
  #Gets all interpro files with isoform root ex) "F28H6" get F28H6.1a and F28H6.1b

  #Search env for files with isoform root
  isoform_files <- ls(pos = 1 , pattern = isoform_root)

  lot <- mget(envir = as.environment(1) , isoform_files)
  df <- bind_rows(lot)

  return(df)
  }
akt2_interpro<- gene_interpro("F28H6")
# daf16_interpro <- gene_interpro("R13H8")
daf2_interpro <- gene_interpro("Y55D5")

# isoform_root <- "F28H6"
# search_pattern_t <- paste(isoform_root)
# ls(pattern = search_pattern_t)
#
# gene_interpro("F28H6")

#
# df <- mget(akt2_files)
# bind_rows(df)



load_gene <- function(file) {
  file_name <- print(file)
  gene_name <- str_extract(file_name, "[a-z]+\\d+" )
  print(gene_name)
  file_read <- data.table::fread(file)
  assign(glue("{gene_name}"), file_read, pos = 1 )
  }


#Loop over all gene flat files in Dir
for (g in gene_files){
  load_gene(g)}



##Test Joins ##

#Filter down gene file to only deleterious variants
#Add To laoading gene file





akt2_hi <- akt2 %>% filter(VARIANT_IMPACT == "HIGH") %>%
  filter(AMINO_ACID_CHANGE != "") %>% #Remove Non-protein coding variants
  mutate(AA_POS = sapply(AMINO_ACID_CHANGE, pull_aa_pos)) %>% #Extract First Altered AA POS
  convert(hablar::num(AA_POS)) %>% #Convert AA POS from char to num
  mutate(AA_POS2 = AA_POS) %>% #Create duplicate AA pos column so there is an interval to join by
  convert(hablar::num(AA_POS2)) %>% #Convert AA POS from char to num
  mutate(MAIN_ISO = sapply(TRANSCRIPT, pull_isoform)) #Create main isoform column

# daf16_hi <- daf16 %>% filter(VARIANT_IMPACT == "HIGH") %>%
#   filter(AMINO_ACID_CHANGE != "") %>% #Remove Non-protein coding variants
#   mutate(AA_POS = sapply(AMINO_ACID_CHANGE, pull_aa_pos)) %>% #Extract First Altered AA POS
#   convert(hablar::num(AA_POS)) %>% #Convert AA POS from char to num
#   mutate(AA_POS2 = AA_POS) %>% #Create duplicate AA pos column so there is an interval to join by
#   convert(hablar::num(AA_POS2)) %>% #Convert AA POS from char to num
#   mutate(MAIN_ISO = sapply(TRANSCRIPT, pull_isoform)) #Create main isoform column

# daf18_hi <- daf18 %>% filter(VARIANT_IMPACT == "HIGH") %>%
#   filter(AMINO_ACID_CHANGE != "") %>% #Remove Non-protein coding variants
#   mutate(AA_POS = sapply(AMINO_ACID_CHANGE, pull_aa_pos)) %>% #Extract First Altered AA POS
#   convert(hablar::num(AA_POS)) %>% #Convert AA POS from char to num
#   mutate(AA_POS2 = AA_POS) %>% #Create duplicate AA pos column so there is an interval to join by
#   convert(hablar::num(AA_POS2)) %>% #Convert AA POS from char to num
#   mutate(MAIN_ISO = sapply(TRANSCRIPT, pull_isoform)) #Create main isoform column

daf2_hi <- daf2 %>% filter(VARIANT_IMPACT == "HIGH") %>%
  filter(AMINO_ACID_CHANGE != "") %>% #Remove Non-protein coding variants
  mutate(AA_POS = sapply(AMINO_ACID_CHANGE, pull_aa_pos)) %>% #Extract First Altered AA POS
  convert(hablar::num(AA_POS)) %>% #Convert AA POS from char to num
  mutate(AA_POS2 = AA_POS) %>% #Create duplicate AA pos column so there is an interval to join by
  convert(hablar::num(AA_POS2)) %>% #Convert AA POS from char to num
  mutate(MAIN_ISO = sapply(TRANSCRIPT, pull_isoform)) #Create main isoform column


#mutate(AA_POS = sapply(AMINO_ACID_CHANGE, pull_aa_pos))
  group_by(POS, REF, ALT, CONSEQUENCE, TRANSCRIPT) %>%
  nest()
  mutate(POS2 = POS)

test_var <- akt2_hi[1, "TRANSCRIPT"] %>% as.character()

pull_aa_pos <- function(x){
 str_extract(x, one_or_more(ascii_digit())) #Newer version of rebus
  }

pull_isoform <- function(x){
  toupper(str_extract(x, one_or_more(ALNUM) %R% "." %R% one_or_more(ALNUM)))
}


#Loop to Join based on transcript


#Join is creating duplicates?
#Duplicates are created becuase of transcripts that vary in UTRs

akt2_join <- genome_join(akt2_hi, akt2_interpro ,
            by = c(
              "MAIN_ISO" = "ISOFORM",
              "AA_POS" = "START",
              "AA_POS2" = "STOP"),
              mode = "left")
#
# daf16_join <- genome_join(daf16_hi, daf16_interpro ,
#                       by = c(
#                         "MAIN_ISO" = "ISOFORM",
#                         "AA_POS" = "START",
#                         "AA_POS2" = "STOP"),
#                       mode = "left")
# daf18_join <- genome_join(daf18_hi, T07A9.6,
#                           by = c(
#                             "MAIN_ISO" = "ISOFORM",
#                             "AA_POS" = "START",
#                             "AA_POS2" = "STOP"),
#                           mode = "left")
daf2_join <- genome_join(daf2_hi, daf2_interpro,
                         by = c(
                           "MAIN_ISO" = "ISOFORM",
                           "AA_POS" = "START",
                           "AA_POS2" = "STOP"),
                         mode = "left")

### Variation in Domains ###
###########################

#Filter START is NA

var_in_domains <- function(df){
  df %>% filter(! is.na(STOP)) %>%
    group_by(POS, REF, ALT, CONSEQUENCE,AMINO_ACID_CHANGE, BLOSUM, Grantham, MAIN_ISO, ANNOTATION, ANNOTATION_ID) %>%
    nest()

  #Add Strains ?
}

akt2_domain_var <- var_in_domains(akt2_join)
# daf16_domain_var <- var_in_domains(daf16_join) #No variants in key functional domains
# daf18_domain_var <- var_in_domains(daf18_join)
daf2_domain_var <- var_in_domains(daf2_join)

#Var By Domain##


akt2_domain_var %>% filter()




#Remove Rows that are blank after all transcript files are joined


#Select Key Columns & Remove variants that don't have an interpro annotation
#
# clean_join <- join_1 %>%
#   select(CHROM, POS, REF, ALT,
#          CONSEQUENCE, TRANSCRIPT,MAIN_ISO, Strains,
#          AMINO_ACID_CHANGE , BLOSUM,
#          Grantham, Percent_Protein, SIGNATURE,
#          SCORE , ANNOTATION) %>%
#   group_by(CHROM, POS, REF, ALT,
#            CONSEQUENCE,MAIN_ISO, Strains,
#            AMINO_ACID_CHANGE , BLOSUM,
#            Grantham, Percent_Protein, SIGNATURE,
#            SCORE , ANNOTATION) %>%
#   nest()



#Variants that are paired variants

#Variants that are in paired strains

#Pulling from older scripts



#Can I make it go thorught all of the interpro data that is loaded into the env

#When Joining need to have a way to join non standard UTR regions
#Create a composite key,
  #str_extract every thing before the last period
  #Make key upper case


### Paired Variants Analysis ###
################################


#Not based off of alleles atm
daf2_paired_var <- c( "3028693",
                      "3028637",
                      "3023143",
                      "3019596",
                      "3013429",
                      "3010466",
                      "3008844",
                      "3007199",
											"3008718" 
                      "3005943",
                      "3005900",
                      "3005885",
                      "3005783",
                      "2999982",
                      "2997182",
                      "2997070",
                      "2995992")


akt2_paired_var <- c( "14151475",
                      "14151123",
                      "14149302",
                      "14149299",
                      "14149254",
                      "14147803"
)



paired_daf2<- daf2_domain_var %>% filter(POS %in% daf2_paired_var)

paired_akt2<- akt2_domain_var %>% filter(POS %in% akt2_paired_var)


akt2_paired_var <- c("")




### Deeper Domain Analysis ###

#Daf2 Fibronectin#

#Isoform A

daf2A_fn<- daf2_domain_var %>% filter(MAIN_ISO == "Y55D5A.5A" & ANNOTATION_ID == "IPR003961") %>%
  select(-data) %>%
  unique()

#dafBC<- daf2_domain_var %>% filter(MAIN_ISO == "Y55D5A.5B" & ANNOTATION_ID == "IPR003961")
daf2C_fn<- daf2_domain_var %>% filter(MAIN_ISO == "Y55D5A.5C" & ANNOTATION_ID == "IPR003961")
daf2D_fn <- daf2_domain_var %>% filter(MAIN_ISO == "Y55D5A.5D" & ANNOTATION_ID == "IPR003961")
daf2E_fn <- daf2_domain_var %>% filter(MAIN_ISO == "Y55D5A.5E" & ANNOTATION_ID == "IPR003961")
daf2F_fn <- daf2_domain_var %>% filter(MAIN_ISO == "Y55D5A.5F" & ANNOTATION_ID == "IPR003961")

daf2_domain_var %>% filter(MAIN_ISO == "Y55D5A.5F") %>%
  filter(ANNOTATION_ID == "IPR003961")


#Predicted Daf2 Domain regionfn
fn_A <- daf2_interpro %>% filter(ISOFORM == "Y55D5A.5A" & ANNOTATION_ID == "IPR003961")
